### ISSUE TITLE
Incorrect triangulation of polygons with many holes

### ISSUE THREAD

**Issue by TonyCrane** [3b1b/manim#1331]
I tried to use `Text` to write Chinese, but there were many problems. 
For example, the simplest one is to write the word "回". But the result is this (contains unnecessary connections):
```python
class TextBug(Scene):
    def construct(self):
        txt = Text("回", font="庞门正道标题体", height=FRAME_HEIGHT-1)
        self.add(txt)
```
![result](https://user-images.githubusercontent.com/44120331/106481642-98e06d80-64e7-11eb-8ed7-fe96119374f2.png)
I lowered its opacity and showed the point set and its indices. The result is this: 
```python
class TextBug(Scene):
    def construct(self):
        txt = Text("回", font="庞门正道标题体", height=FRAME_HEIGHT-1)
        txt.set_fill(opacity=0.3)
        self.add(txt)
        points = txt[0].get_points()
        ind = index_labels(points).set_color(PURPLE).set_stroke(width=1)
        self.add(ind)
```
![TextBug](https://user-images.githubusercontent.com/44120331/106481680-a269d580-64e7-11eb-8b5d-e56e37d5cae9.png)
The position and order of the point set are correct. So I suspect that there may be an error in the triangulation, so I displayed the inner triangles in the triangulation, and the result is this: 
```python
class TextBug(Scene):
    def construct(self):
        txt = Text("回", font="庞门正道标题体", height=FRAME_HEIGHT-1)
        txt.set_fill(opacity=0.3)
        self.add(txt)
        points = txt[0].get_points()
        ind = index_labels(points).set_color(PURPLE).set_stroke(width=1)
        self.add(ind)
        tri = txt[0].get_triangulation()
        tri_indices = tri[len(points):]
        a = tri_indices[0::3]
        b = tri_indices[1::3]
        c = tri_indices[2::3]
        for i, _ in enumerate(a):
            self.add(Polygon(points[a[i]], points[b[i]], points[c[i]],\
            stroke_color=WHITE, stroke_width=4, fill_color=ORANGE, fill_opacity=0.5))
```
![TextTri](https://user-images.githubusercontent.com/44120331/106482424-68e59a00-64e8-11eb-80b9-c69e4d0d37e3.png)
Basically it is certain that there is a problem, but I don't know how to solve it. 

---
In addition, this is the svg file of this Chinese character, you can use `SVGMobject` to test it.
[d17bf5c77d8120cd.txt](https://github.com/3b1b/manim/files/5904912/d17bf5c77d8120cd.txt) (change its extension to `.svg`, because GitHub does not support uploading svg files here)
And it this way, not only `Text` will have problems, but `SVGMobject` and even other mobjects may also have such problems.
So, @3b1b please take a look at this.

**Comment by 3b1b** [comment]
The best solution to this would be to find a replacement for the current "earclip_triangulation" function.  Unfortunately, this does not seem to work well for polygons with too many holes.  I have previously looked for a more robust library in python, and if you can find one do let me know.

I can take a look at this example and try to fix it more directly, but I'm hesitant to make any promises.

**Comment by 3b1b** [comment]
One solution here would be to find a more robust polygon triangulation library written in C, not python, and use that.

**Comment by TonyCrane** [comment]
I just tried another simpler Scene:
```python
def get_tri(mob):
    tri = mob.get_triangulation()
    points = mob.get_points()
    tri_indices = tri[len(points):]
    a = tri_indices[0::3]
    b = tri_indices[1::3]
    c = tri_indices[2::3]
    tris = VGroup()
    for i, _ in enumerate(a):
        tris.add(Polygon(points[a[i]], points[b[i]], points[c[i]],\
        stroke_color=WHITE, stroke_width=4, fill_color=ORANGE, fill_opacity=0.5))
    return tris

def get_dots(mob):
    points = mob.get_points()
    dots = VGroup()
    for point in points:
        dots.add(SmallDot(point, color=PURPLE))
    return dots

class SquaresTri(Scene):
    def construct(self):
        squares = VGroup(*[
            Square(i) for i in np.arange(7, 0, -2)
        ])
        self.add(squares)
        square = VMobject()
        square.set_fill(opacity=0.3)
        for index, mob in enumerate(squares):
            if index % 2:
                square.append_points(mob.get_points()[::-1])
            else:
                square.append_points(mob.get_points())
        self.add(square)
        self.add(get_tri(square), get_dots(square))
```
The expected result is like (made in cairo-backend version): 
![image](https://user-images.githubusercontent.com/44120331/106545271-0a4e0980-6544-11eb-960a-7e6cabaa9dd7.png)
But the code now give me this:
![image](https://user-images.githubusercontent.com/44120331/106545060-a297be80-6543-11eb-9489-15ac5ee2c575.png)
And if I try more squares, that is to change `np.arange(7, 0, -2)` to `np.arange(7, 0, -1)`, it will give me this error:
```python
Traceback (most recent call last):
  File "manim.py", line 5, in <module>
    manimlib.main()
  File "...\manimlib\__init__.py", line 12, in main
    scene.run()
  File "...\manimlib\scene\scene.py", line 76, in run
    self.tear_down()
  File "...\manimlib\scene\scene.py", line 95, in tear_down
    self.interact()
.........
  File "...\manimlib\mobject\types\vectorized_mobject.py", line 800, in get_triangulation      
    inner_tri_indices = inner_vert_indices[earclip_triangulation(inner_verts, rings)]
  File "...\manimlib\utils\space_ops.py", line 364, in earclip_triangulation
    j = int(norms[norms[:, 1].argmin()][0])
IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed
```
So there may exactly be something wrong in `earclip_triangulation`.

**Comment by BenEcon** [comment]
This may be helpful: https://github.com/SebLague/Ear-Clipping-Triangulation/tree/master

**Comment by naveen521kk** [comment]
Am I wrong in saying https://trimsh.org/ would handle this?

**Comment by 3b1b** [comment]
Okay, I think I may have fixed it with [this commit](https://github.com/3b1b/manim/commit/729dae1376fa72500efe2355a28e80c59dc7dee6).  Let me know if that works for you.

**Comment by 3b1b** [comment]
> Am I wrong in saying https://trimsh.org/ would handle this?

Maybe I'm missing something, but it looks like trimesh is useful for working with and manipulating meshes of 3d objects once you already have them, but I'm not seeing where it would be useful for finding the triangulation of a 2d polygon with many holes.

**Comment by BenEcon** [comment]
> Okay, I think I may have fixed it with [this commit](https://github.com/3b1b/manim/commit/729dae1376fa72500efe2355a28e80c59dc7dee6). Let me know if that works for you.

Great, it works now.

**Comment by TonyCrane** [comment]
The aforementioned `SquaresTri` scene has been able to run normally. And some Chinese can also be displayed normally:
![{732I4OG4H0KQF40@1KZY6R](https://user-images.githubusercontent.com/44120331/106679673-26ab7e00-65f8-11eb-8f89-16761af88920.png)
![_9NUQASJ(@$OFWM@KKDHZ61](https://user-images.githubusercontent.com/44120331/106679678-29a66e80-65f8-11eb-9697-9a429a7d3d32.png)
But the "回" mentioned at the beginning and the svg file still cannot be displayed correctly:
![image](https://user-images.githubusercontent.com/44120331/106679722-3b881180-65f8-11eb-8eb6-07d92e0290c7.png)
![5C_QQ HGO((D%_WZ15S~RRX](https://user-images.githubusercontent.com/44120331/106679731-404cc580-65f8-11eb-8637-d431e30e1f24.png)
I guess it’s because the points of the svg with the word "回" are concentrated first on the inner circle and then on the outer circle, causing the outer circle to cover the inner circle:
![V96Z157K)_FXUG}0WJGMO_S](https://user-images.githubusercontent.com/44120331/106679760-4b9ff100-65f8-11eb-924a-98c20cd020dd.png)

**Comment by 3b1b** [comment]
Can you attach the relevant svg?  When I render that character in other fonts it looks fine, but I don't currently have the font from the example above installed.

**Comment by TonyCrane** [comment]
Here: [d17bf5c77d8120cd.txt](https://github.com/3b1b/manim/files/5904912/d17bf5c77d8120cd.txt) (change its extension to `.svg`, because GitHub does not support uploading `svg` files here)
(I sent this svg at the front of this issue, but it may not be obvious)

**Comment by 3b1b** [comment]
Okay, I'll look into it.

**Comment by 3b1b** [comment]
Alright, I believe [this change](https://github.com/3b1b/manim/commit/c8f4c4972042ae54f2cf4016f9a8ab502f8ee682) should actually fix it in a more robust way.  It works on the svg you passed, and it seems to work on all the Chinese text I try, let me know how it looks @TonyCrane.

**Comment by TonyCrane** [comment]
![%8 4}`W23PWCZ@G0HQ}(3IV](https://user-images.githubusercontent.com/44120331/106822256-c54be380-66b9-11eb-8e26-00b33cc55578.png)
That's great! And I also tried some other characters and even a whole passage, they all look good.

**Comment by TonyCrane** [comment]
But there may be still a bug. When I tried to run the `SquaresTri` Scene, it can't display properly, which can display correctly before this changes.
```python
class SquaresTri(Scene):
    def construct(self):
        squares = VGroup(*[
            Square(i) for i in np.arange(7, 0, -1)
        ])
        self.add(squares)
        square = VMobject()
        square.set_fill(opacity=0.3)
        for index, mob in enumerate(squares):
            if index % 2:
                square.append_points(mob.get_points()[::-1])
            else:
                square.append_points(mob.get_points())
        self.add(square)
```
![V_XEC2I`JWDY}8`1RT@@L_Q](https://user-images.githubusercontent.com/44120331/106822832-cdf0e980-66ba-11eb-857d-230e435443b3.png)

**Comment by 3b1b** [comment]
Fascinating.  Unless I am mistaken, _believe_ this is an error with the mapbox_earcut library, and not with the wrapper I wrote around it.

In some sense, this is an "unstable" exception.  What I mean that is that if you perturb the squares slightly, e.g.

`        squares = VGroup(*[
            Square(i).shift(0.1 * i * RIGHT) for i in np.arange(7, 0, -1)
        ])
`
then it works properly.  Also, if you call square.insert_n_curves(10), it will render correctly.

The wrapper is just there to turn a polygon with holes into a continuous polygon by connecting the holes, analogous to how the limit of a [keyhole contour](https://math.stackexchange.com/questions/2245314/complex-keyhole-contour-integral) is a discontinuous contour with a hole in it.  The way it connects up indices in the perturbed example is the same as how it does in the faulty centered example, which leads me to believe it's not an issue with that connecting process.  From there, it's up to the mapbox_earcut library to fill in the single continuous polygon correctly.

Perhaps this means the ultimate solution really does lie with finding a better triangulation library.

If it's alright, I'll go ahead and close this issue since it seems like the exception involving holes within holes within holes may be pathological enough not to come up in practice.  But of course, feel free to reopen it if it does.

**Comment by BenEcon** [comment]
The Chinese character shown in svg is not correctly displayed. 
[54bc7458065788e9.txt](https://github.com/3b1b/manim/files/5922454/54bc7458065788e9.txt)
Change txt to svg.

**Comment by BenEcon** [comment]
> Fascinating. Unless I am mistaken, I _believe_ this is an error with the mapbox_earcut library, and not with the wrapper I wrote around it.
> 
> In some sense, this is an "unstable" exception. What I mean that is that if you perturb the squares slightly, e.g.
> 
> ` squares = VGroup(*[ Square(i).shift(0.1 * i * RIGHT) for i in np.arange(7, 0, -1) ])`
> then it works properly. Also, if you call square.insert_n_curves(10), it will render correctly.
> 
> The wrapper is just there to turn a polygon with holes into a continuous polygon by connecting the holes, analogous to how the limit of a [keyhole contour](https://math.stackexchange.com/questions/2245314/complex-keyhole-contour-integral) is a discontinuous contour with a hole in it. The way it connects up indices in the perturbed example is the same as how it does in the faulty centered example, which leads me to believe it's not an issue with that connecting process. From there, it's up to the mapbox_earcut library to fill in the single continuous polygon correctly.
> 
> Perhaps this means the ultimate solution really does lie with finding a better triangulation library.
> 
> If it's alright, I'll go ahead and close this issue since it seems like the exception involving holes within holes within holes may be pathological enough not to come up in practice. But of course, feel free to reopen it if it does.

The Chinese character shown in svg is not correctly displayed.
54bc7458065788e9.txt
Change txt to svg.

**Comment by 3b1b** [comment]
Okay, maybe not as pathological as I suspect.  I'll look into if this can be fixed without changing the triangulation library, otherwise, that's probably what we'll need to do.

**Comment by BenEcon** [comment]
> Okay, maybe not as pathological as I suspect. I'll look into if this can be fixed without changing the triangulation library, otherwise, that's probably what we'll need to do.

There are two versions about this issue you offered before this one, they all work. But this one, it fails.

**Comment by 3b1b** [comment]
Man, this is getting hackier by the minute.  I've added done more tweak, which seems to fix this character and the others I try, but clearly the game of small tweaks to fix edge cases suggests a deeper fix may still be needed.

![image](https://user-images.githubusercontent.com/11601040/106833670-cc86d900-6648-11eb-9fd2-fb1e215a1578.png)

**Comment by BenEcon** [comment]
> Man, this is getting hackier by the minute. I've added one more tweak, which seems to fix this character and the others I try, but clearly, the game of small tweaks to fix edge cases suggests a deeper fix may still be needed.
> 
> ![image](https://user-images.githubusercontent.com/11601040/106833670-cc86d900-6648-11eb-9fd2-fb1e215a1578.png)

Great, at least for 99.9% scenarios, it is enough. If possible, I think it is a great idea to have a video lecture about earcut algorithm and how to cut and paste different parts by some examples like the aforementioned to work with earcut algorithm, I think it will be interesting and very useful.
